docs/source/settings.rst
Settings
========

.. py:data:: lua_ls_project_root
   :type: str

   Path to a directory with ``.luarc.json`` file, relative to the location
   of ``conf.py``. Lua analyzer will be launched from here.

.. py:data:: lua_ls_project_directories
   :type: list[str]

   List of directories where lua analyzer will run,
   relative to :py:data:`lua_ls_project_root`.

   By default, consists of a single :py:data:`lua_ls_project_root`.

.. py:data:: lua_ls_backend
   :type: str

   Controls which lua analyzer is used. Can be either ``"emmylua"``, ``"luals"``, or ``"disable"``.
   When set to ``"disable"``, no Lua files are analyzed, all language server related settings
   are ignored, and :doc:`autodoc` does not work.

.. py:data:: lua_ls_auto_install
   :type: bool

   Controls whether Sphinx-LuaLs should try downloading lua analyzer (LuaLs or EmmyLua)
   from github if it isn't installed already. This setting is enabled by default.

.. py:data:: lua_ls_auto_install_location
   :type: str

   Controls where the lua analyzer will be installed. By default,
   Sphinx-LuaLs uses a folder in the temporary directory provided by the os.
   For unix, it is ``/tmp/python_lua_ls_cache``.

.. py:data:: lua_ls_min_version
   :type: str

   Controls the minimum version of the used lua analyzer.

   Analyzer version should be greater than or equal to this version.

   For LuaLs, default value is ``3.0.0``, for EmmyLua it's ``0.11.0``.

.. py:data:: lua_ls_max_version
   :type: str | None

   Controls the maximum version of the used lua analyzer.

   Analyzer version should be strictly less than this version.

   For LuaLs, default value is ``4.0.0``, for EmmyLua it's ``2.0.0``.
   Use ``None`` to allow any version.

.. py:data:: lua_ls_skip_versions
   :type: list[str] | None

   List of lua analyzer versions that shouldn't be used.

   For example, if ``lua_ls_skip_versions = ["3.16"]``, language server release
   ``3.16`` (including all ``3.16.x`` releases) will not be used.

   This setting can be used to avoid broken releases without limiting max version
   of the lua analyzer.

   Default is to avoid known broken releases:

   - LuaLs ``3.16.0`` — broken on Linux, fixed in ``3.16.1``.

.. py:data:: lua_ls_default_options
   :type: dict[str, str]

   Default values for directive options. You can override member ordering
   or enable documentation for undocumented or private members from here.
   For example:

   .. code-block:: python

      lua_ls_default_options = {
          # Enable documentation for object's members.
          # Empty string means documenting all members with non-empty description.
          "members": "",
          # Set ordering of automatically generated content to alphabetical.
          "member-order": "alphabetical",
          # And so on...
      }

   If using defaults for the ``:members:``, ``:exclude-members:``, and other
   list options, setting the option on a directive will override the default.
   Instead, to extend the default list with the per-directive option,
   the list may be prepended with a plus sign (``+``), as follows:

   .. code-block:: rst

      .. lua:autoobject:: Noodle
         :members: eat
         :private-members: +_spicy, _garlickly

   Also, the defaults can be disabled per-directive with the negated form,
   ``:no-option:`` as an option of the directive:

   .. code-block:: rst

      .. lua:autoobject:: foo
         :no-undoc-members:

.. py:data:: lua_ls_class_default_function_name
   :type: str

   Allows specifying which class method represents a class constructor.

   Class constructors are documented separately, depending on
   :rst:dir:`lua:autoobject:class-doc-from`
   and :rst:dir:`lua:autoobject:class-signature` options.

.. py:data:: lua_ls_class_default_force_non_colon
   :type: bool

   If ``True``, Sphinx-LuaLs will remove ``self`` from class constructor's signature.

.. py:data:: lua_ls_class_default_force_return_self
   :type: bool

   If ``True``, Sphinx-LuaLs will replace class constructor's return type with ``self``.

.. py:data:: lua_ls_lua_version
   :type: str

   Controls which documentation version is used when linking
   to standard library functions. Does not otherwise affect parsing or generation.

   Can be either ``"5.1"``, ``"5.2"``, ``"5.3"``, ``"5.4"``, or ``"jit"``.

   By default, Sphinx-LuaLs will choose this setting
   based on your ``.emmyrc.json``/``.luarc.json`` file.

.. py:data:: lua_ls_apidoc_roots
   :type: dict[str, str | dict[str, Any]]

   Roots for `apidoc <apidoc.html>`_. Keys are full module names
   that should be generated, and values are directories (relative to the location
   of ``conf.py``) where ``.rst`` files are placed.

   Additionally, you can override other apidoc settings for each root. For this,
   make root's value a dictionary with keys ``path``,
   :py:data:`max_depth <lua_ls_apidoc_max_depth>`,
   :py:data:`options <lua_ls_apidoc_default_options>`,
   :py:data:`ignored_modules <lua_ls_apidoc_ignored_modules>`,
   :py:data:`separate_members <lua_ls_apidoc_separate_members>`,
   and :py:data:`format <lua_ls_apidoc_format>`:

   .. code-block:: python

      lua_ls_apidoc_roots = {
          "moduleName": {
              "path": "moduleDirectory",
              "max_depth": 2,
              "options": {
                  "undoc-members": "",
              },
              "format": "md",
          },
      }

.. py:data:: lua_ls_apidoc_default_options
   :type: dict[str, str]

   Default options for objects documented via apidoc. Override
   :py:data:`lua_ls_default_options`.

.. py:data:: lua_ls_apidoc_max_depth
   :type: int

   Maximum nesting level for files. Submodules that are deeper than this level
   will not get their own file, and instead will be generated inline.

   Default value is ``4``.

.. py:data:: lua_ls_apidoc_separate_members
   :type: bool

   If set to ``True``, module members will be rendered on separate pages.

   .. warning::

      **Windows users**

      This option might not work correctly on case-insensitive file systems.

      It will generate a separate file for every member of a module;
      if there are members that only differ in case (i.e. ``Class`` vs ``class``),
      one of them will overwrite the file for another.

      If you're on Windows, and you experience difficulties because of it,
      `make your source and output directories case-insensitive`_
      and add the following hack to your ``conf.py``:

      .. code-block:: python

         # This evil code forces Python to treat
         # windows filenames as case-sensitive.
         import pathlib
         pathlib.PureWindowsPath._str_normcase = property(str)

.. _make your source and output directories case-insensitive:
   https://learn.microsoft.com/en-us/windows/wsl/case-sensitivity

.. py:data:: lua_ls_apidoc_ignored_modules
   :type: list[str]

   List of full submodule names that should be ignored while generating APIs.
   Submodules can contain :py:mod:`fnmatch` style globs.

   For example, the following setting

   .. code-block:: python

      lua_ls_apidoc_roots = {
          "moduleName": "moduleDirectory",
      }

      lua_ls_apidoc_ignored_modules = [
         "moduleName.submoduleName"
      ]

   will generate API for module ``moduleName``, but will not include
   ``moduleName.submoduleName``.

.. py:data:: lua_ls_apidoc_format
   :type: str

   Format for generated files. Can be either ``"rst"`` or ``"md"``.

.. py:data:: lua_ls_maximum_signature_line_length
   :type: int | None

   Controls maximum width after which long signatures will be wrapped.

   Default value is ``50``, which is suitable for most Sphinx themes.

   Setting this value to ``None`` will cause signature formatter
   to use Sphinx's global setting ``maximum_signature_line_length``.
   If ``maximum_signature_line_length`` is also ``None``,
   signature wrapping will be disabled.

.. py:data:: lua_ls_verbose
   :type: bool

   Prints debug logs from sphinx-lua-ls without raising verbosity level
   of other Sphinx loggers. Set to ``True`` to see object tree discovered
   by lua analyzer.


_sources/settings.rst.txt
Settings
========

.. py:data:: lua_ls_project_root
   :type: str

   Path to a directory with ``.luarc.json`` file, relative to the location
   of ``conf.py``. Lua analyzer will be launched from here.

.. py:data:: lua_ls_project_directories
   :type: list[str]

   List of directories where lua analyzer will run,
   relative to :py:data:`lua_ls_project_root`.

   By default, consists of a single :py:data:`lua_ls_project_root`.

.. py:data:: lua_ls_backend
   :type: str

   Controls which lua analyzer is used. Can be either ``"emmylua"`` or ``"luals"``.

.. py:data:: lua_ls_auto_install
   :type: bool

   Controls whether Sphinx-LuaLs should try downloading lua analyzer (LuaLs or EmmyLua)
   from github if it isn't installed already. This setting is enabled by default.

.. py:data:: lua_ls_auto_install_location
   :type: str

   Controls where the lua analyzer will be installed. By default,
   Sphinx-LuaLs uses a folder in the temporary directory provided by the os.
   For unix, it is ``/tmp/python_lua_ls_cache``.

.. py:data:: lua_ls_min_version
   :type: str

   Controls the minimum version of the used lua analyzer.

   Analyzer version should be greater than or equal to this version.

   For LuaLs, default value is ``3.0.0``, for EmmyLua it's ``0.11.0``.

.. py:data:: lua_ls_max_version
   :type: str | None

   Controls the maximum version of the used lua analyzer.

   Analyzer version should be strictly less than this version.

   For LuaLs, default value is ``4.0.0``, for EmmyLua it's ``2.0.0``.
   Use ``None`` to allow any version.

.. py:data:: lua_ls_default_options
   :type: dict[str, str]

   Default values for directive options. You can override member ordering
   or enable documentation for undocumented or private members from here.
   For example:

   .. code-block:: python

      lua_ls_default_options = {
          # Enable documentation for object's members.
          # Empty string means documenting all members with non-empty description.
          "members": "",
          # Set ordering of automatically generated content to alphabetical.
          "member-order": "alphabetical",
          # And so on...
      }

   If using defaults for the ``:members:``, ``:exclude-members:``, and other
   list options, setting the option on a directive will override the default.
   Instead, to extend the default list with the per-directive option,
   the list may be prepended with a plus sign (``+``), as follows:

   .. code-block:: rst

      .. lua:autoobject:: Noodle
         :members: eat
         :private-members: +_spicy, _garlickly

   Also, the defaults can be disabled per-directive with the negated form,
   ``:no-option:`` as an option of the directive:

   .. code-block:: rst

      .. lua:autoobject:: foo
         :no-undoc-members:

.. py:data:: class_default_function_name
   :type: str

   Allows specifying which class method represents a class constructor.

   Class constructors are documented separately, depending on
   :rst:dir:`lua:autoobject:class-doc-from`
   and :rst:dir:`lua:autoobject:class-signature` options.

   If using EmmyLua as lua analyzer, this option will be inferred from ``.emmyrc.json``.

.. py:data:: class_default_force_non_colon
   :type:

   If ``True``, Sphinx-LuaLs will remove ``self`` from class constructor's signature.

   If using EmmyLua as lua analyzer, this option will be inferred from ``.emmyrc.json``.

.. py:data:: class_default_force_return_self
   :type:

   If ``True``, Sphinx-LuaLs will replace class constructor's return type with ``self``.

   If using EmmyLua as lua analyzer, this option will be inferred from ``.emmyrc.json``.

.. py:data:: lua_ls_lua_version
   :type: str

   Controls which documentation version is used when linking
   to standard library functions. Does not otherwise affect parsing or generation.

   Can be either ``"5.1"``, ``"5.2"``, ``"5.3"``, ``"5.4"``, or ``"jit"``.

   By default, Sphinx-LuaLs will choose this setting
   based on your ``.emmyrc.json``/``.luarc.json`` file.

.. py:data:: lua_ls_apidoc_roots
   :type: dict[str, str | dict[str, Any]]

   Roots for `apidoc <apidoc.html>`_. Keys are full module names
   that should be generated, and values are directories (relative to the location
   of ``conf.py``) where ``.rst`` files are placed.

   Additionally, you can override other apidoc settings for each root. For this,
   make root's value a dictionary with keys ``path``,
   :py:data:`max_depth <lua_ls_apidoc_max_depth>`,
   :py:data:`options <lua_ls_apidoc_default_options>`,
   :py:data:`ignored_modules <lua_ls_apidoc_ignored_modules>`,
   :py:data:`separate_members <lua_ls_apidoc_separate_members>`,
   and :py:data:`format <lua_ls_apidoc_format>`:

   .. code-block:: python

      lua_ls_apidoc_roots = {
          "moduleName": {
              "path": "moduleDirectory",
              "max_depth": 2,
              "options": {
                  "undoc-members": "",
              },
              "format": "md",
          },
      }

.. py:data:: lua_ls_apidoc_default_options
   :type: dict[str, str]

   Default options for objects documented via apidoc. Override
   :py:data:`lua_ls_default_options`.

.. py:data:: lua_ls_apidoc_max_depth
   :type: int

   Maximum nesting level for files. Submodules that are deeper than this level
   will not get their own file, and instead will be generated inline.

   Default value is ``4``.

.. py:data:: lua_ls_apidoc_separate_members
   :type: bool

   If set to ``True``, module members will be rendered on separate pages.

   .. warning::

      **Windows users**

      This option might not work correctly on case-insensitive file systems.

      It will generate a separate file for every member of a module;
      if there are members that only differ in case (i.e. ``Class`` vs ``class``),
      one of them will overwrite the file for another.

      If you're on Windows, and you experience difficulties because of it,
      `make your source and output directories case-insensitive`_
      and add the following hack to your ``conf.py``:

      .. code-block:: python

         # This evil code forces Python to treat
         # windows filenames as case-sensitive.
         import pathlib
         pathlib.PureWindowsPath._str_normcase = property(str)

.. _make your source and output directories case-insensitive:
   https://learn.microsoft.com/en-us/windows/wsl/case-sensitivity

.. py:data:: lua_ls_apidoc_ignored_modules
   :type: list[str]

   List of full submodule names that should be ignored while generating APIs.
   Submodules can contain :py:mod:`fnmatch` style globs.

   For example, the following setting

   .. code-block:: python

      lua_ls_apidoc_roots = {
          "moduleName": "moduleDirectory",
      }

      lua_ls_apidoc_ignored_modules = [
         "moduleName.submoduleName"
      ]

   will generate API for module ``moduleName``, but will not include
   ``moduleName.submoduleName``.

.. py:data:: lua_ls_apidoc_format
   :type: str

   Format for generated files. Can be either ``"rst"`` or ``"md"``.

.. py:data:: lua_ls_maximum_signature_line_length
   :type: int | None

   Controls maximum width after which long signatures will be wrapped.

   Default value is ``50``, which is suitable for most Sphinx themes.

   Setting this value to ``None`` will cause signature formatter
   to use Sphinx's global setting ``maximum_signature_line_length``.
   If ``maximum_signature_line_length`` is also ``None``,
   signature wrapping will be disabled.


CHANGELOG.md
# Changelog

## [Unreleased]

## [3.11.0] - 2026-03-29

- Added config option to print debug logs: `lua_ls_verbose`.
- Improved error messages related to object lookup errors and added a troubleshooting
  guide.

## [3.10.1] - 2026-03-29

- Fiexd broken cross-references on the search page.

## [3.10.0] - 2026-01-29

- Added option to disable running lua-ls (`lua_ls_backend = "disable"`) for cases when
  only manual documentation is required ([#47] by [@pieterlexis]).

## [3.9.0] - 2026-01-23

- Fixed LaTeX build and improved rendering of multiline signatures in LaTeX ([#46]).

## [3.8.1] - 2025-12-08

- Added LuaLs 3.16.0 to the list of known broken releases ([LuaLS#3301]).

## [3.8.0] - 2025-12-03

- Added support for Sphinx 9.

- Added `lua_ls_skip_versions` config option to selectively avoid certain language server
  releases.

## [3.7.0-post1] - 2025-11-14

- Minor documentation tweaks.

## [3.7.0] - 2025-11-14

- Fixed parsing of `@see` annotations with LuaLs backend.

- Fixed parsing of EmmyLua output after they removed `classDefaultCall` setting.

## [3.6.0] - 2025-10-17

- Adjusted styles for links in object signatures to match those produced by python domain.

- Releases are no longer uploaded to test version of PyPi.

## [3.5.0-post1] - 2025-10-03

- Migrated documentation to Read the Docs.

- Moved repository to [sphinx-contrib](https://github.com/sphinx-contrib) organization.

## [3.5.0] - 2025-10-03

- Added `autodata`, `autoattribute`, `autoclass`, and other `auto*` directives.

  They work like `autoobject`, but apply their doctype to the documented object (if
  `!doctype` was set in source code, it shouldn't conflict with the used directive).

  They also allow overriding object's signature, which may be useful when automatically
  generated signature is too long.

- Added `lua_ls_max_version` config option to safeguard against incompatible changes to
  documentation export format.

- Fixed a few more minor bugs in highlighting of Lua types.

## [3.4.0] - 2025-09-28

- **Potential breaking change:** use `confdir` instead of `srcdir` as base path for
  `lua_ls_project_root`.

  Prior to this change, `lua_ls_project_root` was resolved relative to the directory
  containing source `.rst` files. Documentation, however, was saying that it's resolved
  relative to the directory with `conf.py`.

  This is not an issue, because in most projects `conf.py` and source `.rst` files are
  located in the same directory. Still, I've decided to be consistent with other Sphinx
  extensions and use `confdir` instead of `srcdir`.

  This is a breaking change, but I don't believe there are any projects that use separate
  `confdir` and `srcdir` (the only reason to do this is if you're hosting multiple
  documentation sites in the same repo.) For this reason, this change is released as a
  minor version change.

## [3.3.0] - 2025-09-28

- Added an option to extend list options (like `:exclude-members:`) without overriding
  defaults:

  ```rst
  .. lua:autoobject::
     :exclude-members: +foo
  ```

  Also added `:no-*:` options to ignore defaults.

- Improved display of members which use types instead of names, i.e. `[<type>]` ([#19] by
  [@bkoropoff]).

- Added a warning for situations when `lua_ls_project_directories` contains directories
  outside of the current VCS root.

## [3.2.0] - 2025-09-02

- Updated a few dependencies. Most notably, restricted `sphinx` to `<9`.

- Fixed dashes in types being parsed as type names ([#18] by [@bkoropoff]).

- Added normalization for function syntax (`fun() -> T` is converted to `fun(): T`).

- Done some internal refactorings.

## [3.1.0] - 2025-08-21

- Added pygments lexer for Lua that highlights documentation tags.

- Added option to control maximum signature length before wrapping.

- Sphinx's nitpicky mode will no longer emit warnings for cross-references in signatures.

- Fixed generation of URL anchors to avoid duplicates.

- Fixed some edge cases in parsing of type expressions.

## [3.0.0] - 2025-07-26

- **Breaking change:** changed how `apidoc` generates file names to avoid collisions.

- Supported [EmmyLua] as an alternative backend for documentation export.

  EmmyLua was recently re-implemented in Rust, and its new language server provides some
  substantial benefits:

  - it has stronger and more flexible type system,
  - it handles aliases and enums way better than LuaLs,
  - it supports namespaces,
  - it exports way more metadata, including function overloads and class constructors,
  - you don't have to annotate modules with `@class` and `!doctype` anymore.

  I plan to use EmmyLua as the default language server since *v4.0.0*.

- Long object signatures are now broken into multiple lines.

- Supported using arbitrary types as object names, i.e.:

  ```rst
  .. lua:data:: [integer]: string
  ```

- Added `:globals:` flag for the `lua:autoobject` directive. It will allow automatically
  documenting global variables defined in a module.

- The `lua:autoindex` directive now lists globals defined in a module.

- Added support for documenting class constructors:

  ```rst
  .. lua:class:: Foo(a, b, ...)
  ```

- Added directive for enums.

- Supported generic parameters for functions, classes, aliases and enums:

  ```rst
  .. lua:class:: Foo<T>
  ```

- Improved linking to Lua language documentation to take into account whether an item is
  supported for the given Lua version.

- Added the `lua:lua` role to compensate for MySt not supporting default roles.

  If Lua is your primary domain cross-referencing from markdown can be done like this:

  ```md
  Reference to a {lua}`logging.Logger.info`.
  ```

- Added the `lua:other-inherited-members` directive and `:inherited-members-table:` flag
  for the `lua:autoobject` directive.

  These allow listing all members that were inherited by a class but weren't documented
  within the class body (see [#3]).

- Supported markdown output for `apidoc`.


sphinx_lua_ls/config.py
from __future__ import annotations

import dataclasses
import pathlib
import re
import sys
import typing as _t
from dataclasses import dataclass

import sphinx.application
from sphinx.errors import ConfigError
from sphinx.util import logging

import sphinx_lua_ls.autodoc
import sphinx_lua_ls.domain

_logger = logging.getLogger("sphinx_lua_ls")

T = _t.TypeVar("T")
A = _t.ParamSpec("A")


@dataclass
class LuaDomainConfig:
    project_root: pathlib.Path
    backend: _t.Literal["emmylua", "luals", "disable"] = "luals"
    project_directories: list[pathlib.Path] | None = None
    auto_install: bool = True
    auto_install_location: pathlib.Path | None = None
    min_version: str | None = None
    max_version: str | None = None
    skip_versions: list[str] | None = None
    lua_version: str | None = None
    default_options: dict[str, _t.Any] = dataclasses.field(default_factory=dict)
    apidoc_default_options: dict[str, _t.Any] = dataclasses.field(default_factory=dict)
    apidoc_max_depth: int = 4
    apidoc_ignored_modules: set[str] = dataclasses.field(default_factory=set)
    apidoc_format: _t.Literal["rst", "md"] = "rst"
    apidoc_separate_members: bool = False
    apidoc_roots: dict[str, dict[str, _t.Any]] = dataclasses.field(default_factory=dict)
    class_default_function_name: str | None = None
    class_default_force_non_colon: bool = False
    class_default_force_return_self: bool = False
    maximum_signature_line_length: int | None = 50
    verbose: bool = False


def _type(name: str, value, types: _t.Type[T] | tuple[_t.Type[T], ...]) -> T:
    if not isinstance(value, types):
        if not isinstance(types, tuple):
            types = (types,)
        raise ConfigError(
            f"{name} should be {' or '.join(map(str, types))}, got {type(value)} instead"
        )
    return value


def _str_choices(name: str, value, choices: list[str]) -> str:
    value = _type(name, value, str)
    if value not in choices:
        raise ConfigError(
            f"{name} should be one of {', '.join(map(repr, choices))}, got {value!r} instead"
        )
    return value


def _version(name: str, value) -> str:
    _type(name, value, str)
    if not re.match(r"\d+(\.\d+)*", value):
        raise ConfigError(f"incorrect {name}: {value}")
    return value


def _path(name: str, value, root: str | pathlib.Path) -> pathlib.Path:
    _type(name, value, (str, pathlib.Path))
    try:
        return pathlib.Path(root, value).expanduser().resolve()
    except ValueError as e:
        raise ConfigError(f"incorrect lua_ls_project_root: {e}") from None


def _list(
    name: str,
    value,
    checker: _t.Callable[_t.Concatenate[str, object, A], T],
    *args: A.args,
    **kwargs: A.kwargs,
) -> list[T]:
    if value is None:
        value = []
    _type(name, value, list)
    return [checker(f"{name}[{i}]", v, *args, **kwargs) for i, v in enumerate(value)]


def _options(name: str, value) -> dict[str, _t.Any]:
    if value is None:
        value = {}
    _type(name, value, dict)
    new_value = {}
    for key, option in value.items():
        parser = sphinx_lua_ls.autodoc.AutoObjectDirective.option_spec.get(key, None)
        if parser is None:
            raise ConfigError(f"unknown option in {name}: :{key}:")
        if key not in sphinx_lua_ls.domain.GLOBAL_OPTIONS:
            raise ConfigError(
                f"incorrect option in {name}: :{key}: can't be set from config"
            )
        _type(f"{name}[{key!r}]", option, str)
        try:
            new_value[key] = parser(option)
        except Exception as e:
            raise ConfigError(f"incorrect {name}[{key!r}]: {e}") from None
    return new_value


def _api_roots(
    name: str,
    value,
    root: str | pathlib.Path,
    max_depth: int,
    options: dict[str, _t.Any],
    excludes: set[str],
    format: str,
    separate_members: bool,
) -> dict[str, dict[str, _t.Any]]:
    if value is None:
        value = {}
    _type(name, value, dict)
    new_value = {}
    for mod in value:
        api_root = value[mod]
        if isinstance(api_root, str):
            api_root = {"path": _path(f"{name}[{mod!r}]", api_root, root)}
        _type(f"{name}[{mod!r}]", api_root, (dict, str))

        new_api_root = {}
        new_api_root["path"] = _path(
            f"{name}[{mod!r}]['path']", api_root.pop("path", None), root
        )
        if not new_api_root["path"].is_relative_to(root):
            raise ConfigError(
                f"api root {name}[{mod!r}] lays outside of src root: {str(new_api_root['path'])}"
            )
        new_api_root["options"] = options.copy()
        new_api_root["options"].update(
            _options(f"{name}[{mod!r}]['options']", api_root.pop("options", None))
        )
        new_api_root["max_depth"] = _type(
            f"{name}[{mod!r}]['max_depth']", api_root.pop("max_depth", max_depth), int
        )
        new_api_root["ignored_modules"] = _excludes(
            f"{name}[{mod!r}]['ignored_modules']",
            api_root.pop("ignored_modules", excludes),
        )
        new_api_root["format"] = _str_choices(
            f"{name}[{mod!r}]['format']", api_root.pop("format", format), ["rst", "md"]
        )
        new_api_root["separate_members"] = _type(
            f"{name}[{mod!r}]['separate_members']",
            api_root.pop("separate_members", separate_members),
            bool,
        )
        if api_root:
            raise ConfigError(
                f"unknown keys in {name}[{mod!r}]: {', '.join(map(str, api_root))}"
            )
        new_value[mod] = new_api_root
    return new_value


def _excludes(name: str, value) -> set[str]:
    if value is None:
        value = []
    _type(name, value, (list, set))
    if isinstance(value, list):
        for i in range(len(value)):
            _type(f"{name}[{i}]", value[i], str)
    else:
        for v in value:
            _type(f"{name}[{v}]", v, str)
    return set(value)


def set_options(app: sphinx.application.Sphinx):
    config = app.config

    project_root = _path(
        "lua_ls_project_root", config["lua_ls_project_root"] or "", app.confdir
    )

    domain_config = LuaDomainConfig(project_root=project_root)

    if config["lua_ls_backend"] is not None:
        domain_config.backend = _t.cast(
            _t.Literal["emmylua", "luals", "disable"],
            _str_choices(
                "lua_ls_backend",
                config["lua_ls_backend"],
                ["luals", "emmylua", "disable"],
            ),
        )
    else:
        _logger.warning(
            "Sphinx-LuaLs will use EmmyLua as the default language server since v4.0. "
            "To keep using LuaLs, set `lua_ls_backend='luals' in your conf.py`",
            type="lua-ls",
        )

    if config["lua_ls_project_directories"] is not None:
        domain_config.project_directories = _list(
            "lua_ls_project_directories",
            config["lua_ls_project_directories"],
            _path,
            domain_config.project_root,
        )

    if config["lua_ls_auto_install"] is not None:
        domain_config.auto_install = _type(
            "lua_ls_auto_install", config["lua_ls_auto_install"], bool
        )

    if config["lua_ls_auto_install_location"] is not None:
        domain_config.auto_install_location = _path(
            "lua_ls_auto_install_location",
            config["lua_ls_auto_install_location"],
            pathlib.Path("/"),
        )

    if config["lua_ls_min_version"] is not None:
        domain_config.min_version = _version(
            "lua_ls_min_version", config["lua_ls_min_version"]
        )
    if config["lua_ls_max_version"] is not None:
        domain_config.max_version = (
            _version("lua_ls_max_version", config["lua_ls_max_version"])
            if config["lua_ls_max_version"] != "__auto__"
            else "__auto__"
        )
    if config["lua_ls_skip_versions"] is not None:
        domain_config.skip_versions = _list(
            "lua_ls_skip_versions", config["lua_ls_skip_versions"], _version
        )

    if config["lua_ls_lua_version"]:
        domain_config.lua_version = _str_choices(
            "lua_ls_lua_version",
            config["lua_ls_lua_version"],
            ["jit", "5.1", "5.2", "5.3", "5.4", "5.5"],
        )

    domain_config.default_options = _options(
        "lua_ls_default_options", config["lua_ls_default_options"]
    )
    domain_config.apidoc_default_options = domain_config.default_options.copy()
    domain_config.apidoc_default_options.update(
        _options(
            "lua_ls_apidoc_default_options", config["lua_ls_apidoc_default_options"]
        )
    )

    if config["lua_ls_apidoc_max_depth"] is not None:
        domain_config.apidoc_max_depth = _type(
            "lua_ls_apidoc_max_depth", config["lua_ls_apidoc_max_depth"], int
        )

    if config["lua_ls_apidoc_ignored_modules"] is not None:
        domain_config.apidoc_ignored_modules = _excludes(
            "lua_ls_apidoc_ignored_modules", config["lua_ls_apidoc_ignored_modules"]
        )

    if config["lua_ls_apidoc_format"] is not None:
        domain_config.apidoc_format = _t.cast(
            _t.Literal["rst", "md"],
            _str_choices(
                "lua_ls_apidoc_format", config["lua_ls_apidoc_format"], ["rst", "md"]
            ),
        )

    if config["lua_ls_apidoc_separate_members"] is not None:
        domain_config.apidoc_separate_members = _type(
            "lua_ls_apidoc_separate_members",
            config["lua_ls_apidoc_separate_members"],
            bool,
        )

    domain_config.apidoc_roots = _api_roots(
        "lua_ls_apidoc_roots",
        config["lua_ls_apidoc_roots"],
        app.confdir,
        domain_config.apidoc_max_depth,
        domain_config.apidoc_default_options,
        domain_config.apidoc_ignored_modules,
        domain_config.apidoc_format,
        domain_config.apidoc_separate_members,
    )

    if config["lua_ls_class_default_function_name"] is not None:
        domain_config.class_default_function_name = _type(
            "lua_ls_class_default_function_name",
            config["lua_ls_class_default_function_name"],
            str,
        )

    if config["lua_ls_class_default_force_non_colon"] is not None:
        domain_config.class_default_force_non_colon = _type(
            "lua_ls_class_default_force_non_colon",
            config["lua_ls_class_default_force_non_colon"],
            bool,
        )

    if config["lua_ls_class_default_force_return_self"] is not None:
        domain_config.class_default_force_return_self = _type(
            "lua_ls_class_default_force_return_self",
            config["lua_ls_class_default_force_return_self"],
            bool,
        )

    if config["lua_ls_maximum_signature_line_length"] is not None:
        domain_config.maximum_signature_line_length = _type(
            "lua_ls_maximum_signature_line_length",
            config["lua_ls_maximum_signature_line_length"],
            int,
        )
    else:
        domain_config.maximum_signature_line_length = _type(
            "maximum_signature_line_length",
            config["maximum_signature_line_length"],
            (int, type(None)),
        )

    if config["lua_ls_verbose"] is not None:
        domain_config.verbose = _type("lua_ls_verbose", config["lua_ls_verbose"], bool)
        if domain_config.verbose:
            app.verbosity = max(app.verbosity, 1)
            root_logger = logging.getLogger("sphinx_lua_ls")
            root_logger.logger.propagate = False
            handler = logging.NewLineStreamHandler(
                logging.SafeEncodingWriter(sys.stdout)
            )
            handler.setFormatter(logging.ColorizeFormatter())
            root_logger.logger.addHandler(handler)
            root_logger.setLevel("DEBUG")

    domain = _t.cast(sphinx_lua_ls.domain.LuaDomain, app.env.get_domain("lua"))
    domain.config = domain_config
